feat(ai): add abort reason field to stream chunks

EurFelux opened the pull request:
## Background
Issue #11442 requests abort reason visibility on abort chunks so callers can distinguish manual vs timeout aborts when combining signals.

## Summary
- Attach `AbortSignal.reason` to `abort` chunks in `streamText`, and extend stream/UI chunk types and schemas to carry `reason`.
- Add tests covering abort reason propagation and update affected snapshots.
- Add an example under `examples/ai-core/src/stream-text` demonstrating abort reasons with `AbortSignal.any` (manual vs timeout).
- Document the optional `reason` field on abort stream parts in core and UI stream protocol docs.

## Manual Verification
- Ran `pnpm -C /Volumes/CIRCLE/dev/ai/packages/ai exec tsx /Volumes/CIRCLE/dev/ai/examples/ai-core/src/stream-text/openai-compatible-abort-reason.ts` and observed abort reasons for both manual and timeout cases.

## Checklist
- [x] Tests have been added / updated (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] A _patch_ changeset for relevant packages has been added (for bug fixes / features - run `pnpm changeset` in the project root)
- [x] I have reviewed this pull request (self-review)

## Related Issues
Fixes https://github.com/vercel/ai/issues/11442

EurFelux committed:
feat(ai): add abort reason field to stream chunks

Pass through abort reason when provided in abort signals to improve debugging and user feedback. Includes updates to types, schemas, and tests to support the new field.

EurFelux committed:
docs(stream-protocol): add abort part documentation and reason field

Add documentation for the abort part in stream protocol and include optional reason field in stream-text reference

EurFelux committed:
docs(changeset): include abort reason in stream chunks and document field

EurFelux committed:
style: fix indentation in stream-text test file

lgrammel commented on packages/ai/src/generate-text/stream-text.test.ts:
```diff
@@ -14423,6 +14491,7 @@ describe('streamText', () => {
                 "warnings": [],
               },
               {
+                "reason": [AbortError: This operation was aborted],
```
this isn't json serializable. is there way to make the reason a string?

EurFelux commented on packages/ai/src/ui-message-stream/ui-message-chunks.ts:
```diff
@@ -168,6 +168,7 @@ export const uiMessageChunkSchema = lazySchema(() =>
       }),
       z.strictObject({
         type: z.literal('abort'),
+        reason: z.unknown().optional(),
```
You’re right; it should be a serializable value.

EurFelux committed:
fix: change abort reason type from unknown to string

Use getErrorMessage to serialize reason for safe JSON transmission in stream text results

EurFelux committed:
test(stream-text): add test for AbortSignal.timeout() reason serialization

EurFelux committed:
refactor(stream-text): simplify abort reason handling and env vars

Remove redundant error type check for abort reason and streamline environment variable naming for Deepseek API

EurFelux committed:
Revert "test(stream-text): add test for AbortSignal.timeout() reason serialization"

This reverts commit 7cdc0ed6fcb02452023568f77cdddbd8a2ad7ee1.

lgrammel committed:
Merge branch 'main' into feat/abort-reason

EurFelux commented on packages/ai/src/generate-text/stream-text.ts:
```diff
@@ -1011,7 +1011,13 @@ class DefaultStreamTextResult<TOOLS extends ToolSet, OUTPUT extends Output>
+            // Use getErrorMessage to serialize reason (unknown) to string for safe JSON transmission
+            ...(abortSignal?.reason !== undefined
+              ? { reason: getErrorMessage(abortSignal.reason) }
+              : {}),
```
You can’t just accept a string only. Please see https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/reason: by default, the `reason` field is set to a value of type `DOMException`, not a `string`. In fact, you can set it to any value using the `abort` function, which is why I previously designed the field’s type as `unknown`. Given that this is the default behavior, if we only accepted `string`, we would lose the `reason` in most use cases.

EurFelux replied:
I used `getErrorMessage` for the sake of code reuse, since this function can serialize errors of type `unknown`. I’m also aware that the fact that `reason` is likely an `Error` instance isn’t immediately obvious from its name, which might confuse maintainers who aren’t familiar with this behavior. To address this, I added a comment explaining why I’m using this function, but it seems the explanation wasn’t clear enough. I’ll provide a more precise explanation of this behavior in the comment.

EurFelux committed:
docs(generate-text): clarify comment about abort signal reason serialization
